emscripten-simulated-jerry-api/jerry.js

```diff
@@ -0,0 +1,305 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+(function () {
+  var global = (new Function('return this;'))();
+
+  // Define garbage collection handler at highest possible scope as
+  // a pure function. Defining it in an inner scope has a high risk
+  // of accidentally creating retain cycles...
+  function _jerryhandleGarbageCollected(internalProps) {
+    console.log('GCd:', internalProps);
+    Module.ccall(
+        '_jerry_call_native_object_free_callbacks',
+        null, ['number', 'number', 'number', 'number'],
+        [internalProps.nativeInfo, internalProps.nativePtr,
+          internalProps.nativeHandleFreeCb, internalProps.nativeHandle]);
+  };
+
+  var canHaveInternalProps = function (value) {
+    if (value === null) {
+      return false;
+    }
+    var typeStr = typeof value;
+    return (typeStr === 'object' || typeStr === 'function');
+  };
+
+  var hasInternalProps = function (value) {
+    return (canHaveInternalProps(value) && __jerry._jerryInternalPropsWeakMap.has(value));
+  };
+
+  global.__jerry = {
+    // Define a WeakMap to maintain the map between JS obj and its internal props
+    _jerryInternalPropsWeakMap: {},
+
+    // _jerryToHostValueMap is a 'map' with entry objects that represent what
+    // values the native side is holding a reference to (reference count is positive).
+    // The _jerryToHostValueMap which will allow us to store and retrieve javascript
+    // values from a jerry_value_t, and perform refcounts on values that we still need an
+    // internal reference to, and avoid them from being garbage collected.
+    _jerryToHostValueMap: {},
+
+    // Counter for the next jerry_value_t value.
+    _nextObjectRef: 1,
+
+    _findEntryByHostValue: function (value) {
+      if (Number.isNaN(value)) {
+        // Special case to find NaN
+        for (var jerry_val in this._jerryToHostValueMap) {
+          var entry = this._jerryToHostValueMap[jerry_val];
+          if (Number.isNaN(entry.value)) {
+            return entry;
+          }
+        }
+      } else {
+        for (var jerry_val in this._jerryToHostValueMap) {
+          var entry = this._jerryToHostValueMap[jerry_val];
+          if (entry.value === value) {
+            return entry;
+          }
+        }
+      }
+      return undefined;
+    },
+
+    _findJerryValueByHostValue: function (value) {
+      var entry = _findEntryByHostValue(value);
+      return entry ? entry.jerry_value : 0;
+    },
+
+    _getEntry: function (jerry_value) {
+      var entry = this._jerryToHostValueMap[jerry_value];
+      if (!entry) {
+        console.log(new Error('Entry at ' + jerry_value + ' does not exist').stack);
+        throw new Error('Entry at ' + jerry_value + ' does not exist');
+      }
+      return entry;
+    },
+
+    _registerHostValue: function (value, pre_existing_jerry_value) {
+      var jerry_value = pre_existing_jerry_value || this._nextObjectRef++;
```

Comment:
Let's check here that `this._nextObjectRef` doesn't overflow the `uint32_t` type that is used on the C side, just throw an `Error` for now. Later we can add some code to find an unused `jerry_value_t`.